IBLT fuzzing

awemany opened the pull request:
Hey George,

here's a couple changes I did to the IBLT for fuzzing the IBLT's API as well as the simple de-/serialization of the graphene data structures. If I find the time and if it makes sense (I have to check the details), I might add another such test for CGrapheneSet.

With these changes, I wasn't able to break CIblt yet using the `api_iblt` test case.

I am not sure about the fixes I did. Maybe you want to return the opposite booleans somewhere. Please check.

awemany committed:
Fuzz graphene data structures that can appear on the wire

awemany committed:
Fuzz CIBlt's API

Also add a isModified() accessor to CIblt to be able to properly
use its interface.

awemany committed:
iblt: Always initialize HashTableEntry to zero

awemany committed:
iblt: Make robust against n_hash == 0 or hashTable.size() == 0

awemany committed:
iblt: Allow to add vectors of any size

bissias commented on src/iblt.cpp:
```diff
@@ -115,7 +115,7 @@ void CIblt::resize(size_t _expectedNumEntries, size_t _valueSize)
 void CIblt::_insert(int plusOrMinus, uint64_t k, const std::vector<uint8_t> v)
 {
-    assert(v.size() == valueSize);
+    // assert(v.size() == valueSize);
```
To allow variable valueSize, I think we would need to make some additional changes, mostly to maintain support of IBLT subtraction.

1) valueSize itself should be entirely removed since it becomes unused.
2) We should create a resize method for HashTableEntry that can increase valueSum to a desired size.
3) Something like CIblt::resize(other) should be created, which resizes each HashTableEntry to the corresponding size in other.

There could be better ways of doing this, but basically we need to support the workflow where IBLT I2 is instantiated and then resized to identically match existing IBLT I1. This is necessary so that we can later perform cell-by-cell subtraction: I1 - I2.

awemany replied:
Hmm, to me it looks like valueSum is *always* auto-extended whenever there's a problem with length-mismatch, or am I mistaken?

bissias replied:
Ah I see. IBLT subtraction will auto-extend valueSum as well. So I agree this should work, although I did not actually test it.

awemany replied:
The `assert` is a danger in its own right, so this is why I think it should be removed. But I'll ask Andrew and the others what they think.

awemany committed:
fuzzer: Very simple test of the CGrapheneSet API

bissias commented on src/iblt.cpp:
```diff
@@ -117,9 +117,14 @@ void CIblt::_insert(int plusOrMinus, uint64_t k, const std::vector<uint8_t> v)
 {
     // assert(v.size() == valueSize);
 
+    if (!n_hash)
```
Since n_hash is only set to a variable quantity in resize, why not throw an assertion error or raise an exception there if n_hash is 0? Really the IBLT won't work at all without even a single hash function.

awemany replied:
Yes, I am fine with either. I just want to make sure that the IBLT has no state where it will create crashing DIV-0 errors. When going the exception route, one should make sure that all IBLTs that are accessed after they have been deserialized are only handled in sections that are properly sitting in `try { ... } catch (std::exception) ..` blocks. I did not check whether that is the case yet(?)

This applies similarly to your comment below as well.

bissias commented on src/iblt.cpp:
```diff
@@ -117,9 +117,14 @@ void CIblt::_insert(int plusOrMinus, uint64_t k, const std::vector<uint8_t> v)
+    if (!n_hash)
+        return;
+    size_t bucketsPerHash = hashTable.size() / n_hash;
+    if (!bucketsPerHash)
```
Similar to n_hash, hashTable is only resized in resize and reset. Why don't we raise an assertion error or runtime exception when it is zero? Again, the IBLT is non-functional with an empty hashTable.

awemany committed:
iblt: Do not allow deserialization with n_hash == 0

awemany committed:
iblt: Remove valueSize variable

awemany committed:
iblt: Use by-reference in insert, erase, addValue

gandrewstone committed:
iblt: Fix vectors passsed by reference

As they are changed during peeling. This is a bare patch from Andrew made into
a commit with an extra comment added each by awemany.
